### PR TITLE
feat(utxo): op_return output so utxo thor swaps carry the swap memo

### PR DESCRIPTION
**Pull request by gomesalexandre**
P0 fund-safety enabling change for UTXO THORChain swaps. Today the app fails closed on UTXO (DOGE/BTC/LTC/BCH) THOR swaps because the fast-vault UTXO tx builder can't carry the swap memo on-chain, and without the memo THORChain can't route the vaulted deposit (funds would be stuck). This adds OP_RETURN memo support so those swaps can ship.

## what
Adds an optional `opReturnData` (UTF-8 THORChain memo) to `BuildUtxoSendOptions`. When present, `buildUtxoSendTx` appends a trailing 0-value OP_RETURN output carrying the memo. Memo-less calls are byte-identical to before.

## how
- `buildOpReturnScript`: `len <= 75` -> `OP_RETURN <len> <data>`; `76..80` -> `OP_RETURN OP_PUSHDATA1 <len> <data>`; `> 80` throws (standard-relay cap).
- The OP_RETURN is appended inside `serializeOutputs`, after recipient + change, so it lands in both `outputsRaw` and `outputsWithCount` that every sighash function consumes.
- Its size (`9 + scriptLen`) is folded into the fee estimate in `buildUtxoSendTx`.
- Threaded through the single `serializeOutputs` call site.

## fund-safety
- **OP_RETURN is in the sighash.** `serializeOutputs` feeds `outputsRaw` (BIP143 `hashOutputs` for BTC/LTC/BCH, Zcash outputs digest) and `outputsWithCount` (legacy sighash for DOGE). So every input signature commits to the memo, it cannot be stripped or altered after signing. Test asserts `signingHashesHex` changes when the memo is present, on both the legacy (DOGE) and BIP143 (BTC) paths.
- **Vault amount preserved.** The recipient (vault) output keeps the FULL amount. OP_RETURN is a separate 0-value output; the extra fee comes out of change/inputs, never by shaving the vault output. Test asserts recipient value is unchanged with vs without the memo, and change shrinks by the added fee.
- **>80 reject.** Non-standard payloads throw rather than build an unbroadcastable tx.

## receipts
- New unit tests `packages/sdk/tests/unit/chains/utxo-opreturn.test.ts` (6 cases, bitcoinjs-lib as decode oracle): exactly one trailing 0-value OP_RETURN with the memo bytes, vault amount preserved, sighash commits to the memo (DOGE + BTC), memo-less byte-identical, both pushdata sizes, >80 reject. All green.
- `packages/sdk` typecheck clean. Full unit suite 2271 passed. Lint clean on touched files.

## paired app change
This is the enabling SDK change. The app PR threads the quote memo into `opReturnData` and flips the current fail-closed guard on UTXO THOR swaps.

🤖 Agent-generated

## Summary by CodeRabbit

* **New Features**
  * Added optional UTF-8 memo support for UTXO send transactions via a trailing `OP_RETURN` output (`0` value).
  * Memo bytes are included in signing, so they can’t be altered after signing; vault/amount handling reflects the added output size.

* **Bug Fixes**
  * Fee sizing now accounts for the serialized `OP_RETURN` output contribution.

* **Tests**
  * Added unit coverage for memo/no-memo behavior, correct `OP_RETURN` push encoding (75/76 and 76–80 boundaries), signing-hash changes, and rejection of empty or >80-byte memos.

### CODE REVIEW

File: packages/sdk/src/chains/utxo/tx.ts [hunk 1]
```diff
@@ -541,20 +541,55 @@ function getSighashZcash(
...
+  if (data.length > MAX_OP_RETURN_DATA) {
+    throw new Error(
+      `OP_RETURN data too large: ${data.length} bytes (max ${MAX_OP_RETURN_DATA}). THORChain swap memos fit within this cap.`
+    )
+  }
+  if (data.length <= 75) {
+    return concat(new Uint8Array([OP_RETURN, data.length]), data)
+  }
+  return concat(new Uint8Array([OP_RETURN, OP_PUSHDATA1, data.length]), data)
+}
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/sdk/src/chains/utxo/tx.ts:577]: Good catch @coderabbitai - tackled in c6229ebc, but stronger than your suggestion. Skipping the output (`opReturnData.length > 0 ? ... : undefined`) is *also* unsafe for a swap: a memo-less deposit to the THOR inbound vault is equally unroutable, so the funds still stick. So `buildOpReturnScript` now **throws** on empty data (same fail-loud posture as the >80-byte cap) - an upstream empty-memo bug dies at build time instead of broadcasting stuck funds. +1 test (`rejects an empty memo`). 7/7 green.

**Review by NeOMakinG — APPROVED** [review]
**APPROVE — fund-safe.** The OP_RETURN output is constructed correctly, is committed into every sighash variant (so the memo can't be stripped/altered post-signing), keeps the vault output whole, and is regression-safe for memo-less txs. I added a boundary test that locks the exact 75/76 direct-push vs OP_PUSHDATA1 selection (`170e63cf`). **Approving this unblocks va#1678** (see below).

## testing approach
Fresh worktree at HEAD; two-lane (Claude load-bearing + Codex gpt-5.4, no findings). Read `buildOpReturnScript`, `serializeOutputs`, the fee/change math in `buildUtxoSendTx`, and all three sighash paths; ran the UTXO suite (linked the built core dist to unblock the monorepo import) and mutation-tested the boundary. bitcoinjs-lib is the decode oracle in the tests.

## verification (all 6 asks)
1. **OP_RETURN construction / 75-76 boundary** — `len<=75 → 6a<len><data>` (direct push; the opcode *is* the length, valid for 1..75), `76..80 → 6a 4c <len><data>` (OP_PUSHDATA1), `>80` and empty both throw. The boundary is exactly right (75 is the max direct-push length). The PR tested 51/78; I added a 75-vs-76 boundary case — **mutation-proven** (flipping `<=75` to `<75` turns it RED).
2. **0-value** — the OP_RETURN output is `writeU64LE(0n) + varint(scriptLen) + script`; value=0, never consumes funds (test asserts `value===0n`).
3. **Byte-exact memo** — `new TextEncoder().encode(opReturnData)` placed verbatim after the length prefix; no truncation/mutation (test round-trips the exact THOR memo via bitcoinjs-lib). This is what va#1678 asserts byte-for-byte against.
4. **Fee/vsize** — `opReturnBytes = 9 + scriptLen` (8 value + 1 varint + scriptLen, correct since scriptLen ≤ 82) is folded into `txSize`; change shrinks by the added fee (test), and `change < 0n → throw insufficient funds` (fail-closed, no underpayment/stuck tx).
5. **Regression (memo-less)** — `opReturnData === undefined → opReturnScript undefined → serializeOutputs skips the append + opReturnBytes=0`, so the unsigned tx and signing hashes are byte-identical (test asserts `unsignedRawHex`/`signingHashesHex` equal + exactly 2 outputs, no `0x6a`).
6. **Change / sighash-commit** — single `serializeOutputs` call site; `numOutputs` counts recipient + change? + opReturn? correctly; and crucially the OP_RETURN is built **inside** `serializeOutputs`, so it feeds `outputsRaw` (BIP143 `hashOutputs` @392 for BTC/LTC/BCH, Zcash outputs digest @515) **and** `outputsWithCount` (legacy sighash @437 for DOGE) — every input signature commits to the memo. Verified all 3 sighash paths consume the memo-carrying outputs.

## blocking / preferably-blocking
None.

## suggestion (found + FIXED here — `170e63cf`, test-only)
Added the exact 75/76 boundary test (the crux of OP_RETURN encoding) since the suite only covered 51/78. Mutation-proven.

## risk
Low. Additive, optional `opReturnData`; memo-less path byte-identical; `>80`/empty/insufficient-funds all fail loud. The changeset is a **minor** bump — additive API, no breaking change (the caret-bump hazard doesn't apply to this PR's own change).

## va#1678 unblock
**Yes.** va#1678 threads the quote memo into `opReturnData` and pins `^2.13.2`; this PR adds that exact API as a minor (→ 2.14.0 on publish), which `^2.13.2` resolves to. Once #984 merges + publishes, va#1678 is unblocked. Recommend va#1678 pin `^2.14.0` explicitly on the follow-up so it can't resolve to a pre-`opReturnData` version (defensive against the general caret-float hazard).

## note (pre-existing, not this PR)
`tsc` flags 2 errors in `packages/core/mpc/.../cardano.ts` + `compileTx.golden.test.ts` (`auxiliaryData` not on `ISigningInput`) — a Cardano/WalletCore type drift untouched by this PR (`utxo/tx.ts` is tsc-clean). The SDK's "Lint and Type Check" CI will surface it independently of this change.

## verdict
APPROVE. OP_RETURN correctly encoded (boundary mutation-proven), 0-value, byte-exact, in every sighash, fee-accounted, memo-less byte-identical, insufficient-funds fail-closed. Unblocks va#1678. UTXO suite green locally; SDK CI running.

---
*Reviewed by the gomes review loop — Claude (load-bearing) + Codex gpt-5.4 (no findings). Boundary test committed to the PR branch, not merged.*
